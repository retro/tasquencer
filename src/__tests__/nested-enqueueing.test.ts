import { Effect } from 'effect';
import { expect, it } from 'vitest';

import { Builder, IdGenerator, Service } from '../index.js';
import { getEnabledTaskNames, makeIdGenerator } from './shared.js';

type WorkflowContext = 'complete' | 'fail' | 'cancel';

const subWorkflowDefinition = Builder.workflow<WorkflowContext>()
  .withName('sub')
  .withParentContext<WorkflowContext>()
  .startCondition('start')
  .task('subT1', (t) =>
    t()
      .withWorkItem((w) =>
        w().onStart(({ startWorkItem, getWorkflowContext }) =>
          Effect.gen(function* ($) {
            const context = yield* $(getWorkflowContext());
            const {
              enqueueCancelWorkItem,
              enqueueCompleteWorkItem,
              enqueueFailWorkItem,
            } = yield* $(startWorkItem());
            if (context === 'cancel') {
              yield* $(enqueueCancelWorkItem());
            } else if (context === 'fail') {
              yield* $(enqueueFailWorkItem());
            } else {
              yield* $(enqueueCompleteWorkItem());
            }
          })
        )
      )
      .onEnable(({ enableTask }) =>
        Effect.gen(function* ($) {
          const { enqueueStartTask } = yield* $(enableTask());
          yield* $(enqueueStartTask());
        })
      )
      .onStart(({ startTask }) =>
        Effect.gen(function* ($) {
          const { initializeWorkItem, enqueueStartWorkItem } = yield* $(
            startTask()
          );
          const { id } = yield* $(initializeWorkItem());
          yield* $(enqueueStartWorkItem(id));
        })
      )
  )
  .endCondition('end')
  .connectCondition('start', (to) => to.task('subT1'))
  .connectTask('subT1', (to) => to.condition('end'));

const workflowDefinition = Builder.workflow<WorkflowContext>()
  .withName('activities')
  .startCondition('start')
  .compositeTask('t1', (ct) =>
    ct()
      .withSubWorkflow(subWorkflowDefinition)
      .onEnable(({ enableTask }) =>
        Effect.gen(function* ($) {
          const { enqueueStartTask } = yield* $(enableTask());
          yield* $(enqueueStartTask());
        })
      )
      .onStart(({ startTask, getWorkflowContext }) =>
        Effect.gen(function* ($) {
          const workflowContext = yield* $(getWorkflowContext());
          const { initializeWorkflow, enqueueStartWorkflow } = yield* $(
            startTask()
          );
          const { id } = yield* $(initializeWorkflow(workflowContext));
          yield* $(enqueueStartWorkflow(id));
        })
      )
  )

  .endCondition('end')
  .connectCondition('start', (to) => to.task('t1'))
  .connectTask('t1', (to) => to.condition('end'));

it('handles nested enqueueing (1)', () => {
  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();

    const service = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow, 'complete')),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* $(service.start());
    const state = yield* $(service.getState());
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set());
    expect(state.workflows[0]?.state).toEqual('completed');
  });

  Effect.runSync(program);
});

it('handles nested enqueueing (2)', () => {
  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();

    const service = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow, 'cancel')),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* $(service.start());
    const state = yield* $(service.getState());
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set());
    expect(state.workflows[0]?.state).toEqual('started');
  });

  Effect.runSync(program);
});

it('handles nested enqueueing (3)', () => {
  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();

    const service = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow, 'fail')),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* $(service.start());
    const state = yield* $(service.getState());
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set());
    expect(state.workflows[0]?.state).toEqual('failed');
  });

  Effect.runSync(program);
});
