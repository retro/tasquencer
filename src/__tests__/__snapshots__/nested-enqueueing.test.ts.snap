// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handles nested enqueueing (1) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "completed",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": "complete",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "completed",
    },
    {
      "context": "complete",
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t1",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "completed",
    },
  ],
}
`;

exports[`handles nested enqueueing (2) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "started",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "canceled",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": "cancel",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
    {
      "context": "cancel",
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t1",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "started",
    },
  ],
}
`;

exports[`handles nested enqueueing (3) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "failed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "failed",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "failed",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": "fail",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "failed",
    },
    {
      "context": "fail",
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t1",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "failed",
    },
  ],
}
`;

exports[`handles simple enqueueing (1) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "completed",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": "complete",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "completed",
    },
    {
      "context": "complete",
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t1",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "completed",
    },
  ],
}
`;

exports[`handles simple enqueueing (2) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "started",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "canceled",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": "cancel",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
    {
      "context": "cancel",
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t1",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "started",
    },
  ],
}
`;

exports[`handles simple enqueueing (3) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "failed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "failed",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "failed",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": "fail",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "failed",
    },
    {
      "context": "fail",
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t1",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "failed",
    },
  ],
}
`;
